pkg/kvstore/store/store.go

```diff
@@ -0,0 +1,429 @@
+	delete(s.localKeys, name)
+	s.mutex.Unlock()
+
+	if ok {
+		if err != nil {
+			s.getLogger().WithError(err).Warning("Unable to delete key in kvstore")
+		}
+
+		key.OnDelete()
+	}
+}
+
+// getLocalKeys returns all local keys
+func (s *SharedStore) getLocalKeys() []Key {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	keys := make([]Key, len(s.localKeys))
+	idx := 0
+	for _, key := range s.localKeys {
+		keys[idx] = key
+		idx++
+	}
+
+	return keys
+}
+
+// getSharedKeys returns all shared keys
+func (s *SharedStore) getSharedKeys() []Key {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	keys := make([]Key, len(s.sharedKeys))
+	idx := 0
+	for _, key := range s.sharedKeys {
+		keys[idx] = key
+		idx++
+	}
+
+	return keys
+}
+
+func (s *SharedStore) getLogger() *logrus.Entry {
+	return log.WithFields(logrus.Fields{
+		"storeName": s.name,
+	})
+}
+
+func (s *SharedStore) updateKey(name string, value []byte) error {
+	s.mutex.Lock()
+	key, ok := s.sharedKeys[name]
+
+	// shared key is seen for the first time
+	if !ok {
+		if s.localKeys[name] != nil {
+			// if local key, reuse key instance
+			s.sharedKeys[name] = s.localKeys[name]
+		} else {
+			// allocate key for keys from collaborators
+			s.sharedKeys[name] = s.conf.KeyCreator()
+		}
+		key = s.sharedKeys[name]
+	}
+	s.mutex.Unlock()
+
+	if err := key.Unmarshal(value); err != nil {
+		return err
+	}
+
+	key.OnUpdate()
+	return nil
+}
+
+func (s *SharedStore) deleteKey(name string) {
+	s.mutex.Lock()
+	existingKey, ok := s.sharedKeys[name]
+	delete(s.sharedKeys, name)
+	s.mutex.Unlock()
+
+	if ok {
+		existingKey.OnDelete()
+	} else {
+		s.getLogger().WithField("key", name).
+			Warning("Unable to find deleted key in local state")
+	}
+}
+
+func (s *SharedStore) listAndStartWatcher() error {
+	listDone := make(chan bool)
+
+	go s.watcher(listDone)
+
+	select {
+	case <-listDone:
+	case <-time.After(listTimeoutDefault):
+		return fmt.Errorf("Time out while retrieving initial list of objects from kvstore")
+	}
+
+	return nil
+}
+
+func (s *SharedStore) watcher(listDone chan bool) {
+	s.kvstoreWatcher = kvstore.ListAndWatch(s.name+"-watcher", s.conf.Prefix, watcherChanSize)
+
+	for {
+		select {
+		case event, ok := <-s.kvstoreWatcher.Events:
+			if !ok {
+				return
+			}
+
+			if event.Typ == kvstore.EventTypeListDone {
+				s.getLogger().Debug("Initial list of objects received from kvstore")
+				close(listDone)
+				continue
+			}
+
+			logger := s.getLogger().WithFields(logrus.Fields{
+				"key":       event.Key,
+				"eventType": event.Typ,
+			})
+
+			logger.Infof("Received key update via kvstore [value %s]", event.Value)
+
+			keyName := strings.TrimPrefix(event.Key, s.conf.Prefix)
+			if keyName[0] == '/' {
+				keyName = keyName[1:]
+			}
+
+			switch event.Typ {
+			case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+				if err := s.updateKey(keyName, event.Value); err != nil {
+					logger.WithError(err).Warningf("Unable to unmarshal store value: %s", event.Value)
```

Comment:
The `event.Value` is a slice of byes so `event.Value` should be cast into a string